src/optimizer/cardinality_estimator.cpp

```diff
@@ -207,17 +168,118 @@ void CardinalityEstimator::AddRelationColumnMapping(LogicalGet *get, idx_t relat
 	}
 }
 
-double CardinalityEstimator::EstimateCardinality(double left_card, double right_card, ColumnBinding left_binding,
-                                                 ColumnBinding right_binding) {
-	idx_t tdom_join_right = GetTDom(right_binding);
-#ifdef DEBUG
-	idx_t tdom_join_left = GetTDom(left_binding);
-	D_ASSERT(tdom_join_left == tdom_join_right);
-#endif
-	D_ASSERT(tdom_join_right != 0);
-	D_ASSERT(tdom_join_right != NumericLimits<idx_t>::Maximum());
-	auto expected_cardinality = (left_card * right_card) / tdom_join_right;
-	return expected_cardinality;
+void UpdateDenom(Subgraph2Denominator *relation_2_denom, RelationsToTDom *relation_to_tdom) {
+	if (relation_to_tdom->has_tdom_hll) {
+		relation_2_denom->denom *= relation_to_tdom->tdom_hll;
+	} else {
+		relation_2_denom->denom *= relation_to_tdom->tdom_no_hll;
+	}
+}
+
+void FindSubgraphMatchAndMerge(Subgraph2Denominator &merge_to, idx_t find_me,
+                               vector<Subgraph2Denominator>::iterator subgraph,
+                               vector<Subgraph2Denominator>::iterator end) {
+	for (; subgraph != end; subgraph++) {
+		if (subgraph->relations.count(find_me) >= 1) {
+			for (auto &relation : subgraph->relations) {
+				merge_to.relations.insert(relation);
+			}
+			subgraph->relations.clear();
+			merge_to.denom *= subgraph->denom;
+			return;
+		}
+	}
+}
+
+double CardinalityEstimator::EstimateCardinalityWithSet(JoinRelationSet *new_set) {
+	double numerator = 1;
+	unordered_set<idx_t> actual_set;
+	for (idx_t i = 0; i < new_set->count; i++) {
+		numerator *= relation_attributes[new_set->relations[i]].cardinality;
+		actual_set.insert(new_set->relations[i]);
+	}
+	vector<Subgraph2Denominator> subgraphs;
```

Comment:
It looks like here you are building the `subgraphs` from the ground-up on every request to `EstimateCardinalityWithSet`. Would it be possible to cache some subgraphs in a map, e.g.:
```c++
unordered_map<JoinRelationSet *, SubgraphDenominator> cached_denominators;`
```

Would that be useful? Or would that make you run into the same symmetry issue as before?